src/UnlitBatchShader.ts

```diff
@@ -48,9 +60,60 @@ export class BatchRawUniformGroup extends RawUniformsGroup {
 
     // [textureIndex, wrapS, wrapT, FREE] vec4 aligned so 4th element is free for use
     this.mapSettings = new Float32Array(this.data, offset, 4 * maxInstances);
+    offset += this.mapSettings.byteLength;
+
+    this.offset = offset;
 
     this.nextIdx = 0;
     this.freed = [];
+    this.meshes = new Array(maxInstances);
+  }
+
+  addMesh(mesh: BatchableMesh, atlas: WebGLAtlasTexture): InstanceID | false {
+    // const meshIndiciesAttribute = geometry.index;
+    const instanceId = this.nextId();
+    const material = mesh.material;
+
+    if (material.map && material.map.image) {
+      const textureId = atlas.addTexture(material.map, tempUvTransform);
+
+      if (textureId === undefined) {
+        console.warn("Mesh could not be batched. Texture atlas full.");
+        this.freeId(instanceId);
+        return false;
+      }
+
+      this.setInstanceUVTransform(instanceId, tempUvTransform);
+      this.setInstanceMapSettings(instanceId, textureId[0], material.map.wrapS, material.map.wrapT);
+    } else {
+      this.setInstanceUVTransform(instanceId, atlas.nullTextureTransform);
+      this.setInstanceMapSettings(instanceId, atlas.nullTextureIndex[0], ClampToEdgeWrapping, ClampToEdgeWrapping);
+    }
+
+    this.setInstanceTransform(instanceId, mesh.matrixWorld);
+    this.setInstanceColor(instanceId, material.color || DEFAULT_COLOR, material.opacity || 1);
+
+    this.meshes[instanceId] = mesh;
+
+    return instanceId;
+  }
+
+  removeMesh(mesh: BatchableMesh, atlas: WebGLAtlasTexture) {
+    const instanceId = this.meshes.indexOf(mesh);
+
+    if (mesh.material.map) {
+      atlas.removeTexture(mesh.material.map);
+    }
+
+    this.setInstanceTransform(instanceId, HIDE_MATRIX);
+    this.freeId(instanceId);
+    this.meshes[instanceId] = null;
+  }
+
+  update(instanceId: number, mesh: BatchableMesh) {
+    // TODO need to account for nested visibility deeper than 1 level
+    this.setInstanceTransform(instanceId, mesh.visible && mesh.parent.visible ? mesh.matrixWorld : HIDE_MATRIX);
```

Comment:
Maybe we should move this into Hubs and expose an API for setting batched mesh visibility? We can use an API like `setMeshVisibility(mesh, visible)` and then in Hubs we can limit the check to 1 level and in Spoke we can call that API directly when changing the visible property on an object.

Reply:
I changed update to just get called once instead of requiring the batch to keep track of instance Ids just to pass it back to the UBO. This seems cleaner and also allows whoever overrides the BatchRawUniformGroup to just take over the whole update a bit more cleanly. In spoke we can just have the update be a noop since spoke already knows better when objects need to be updated. Pretty indifferent to having a `setInstanceVisibility`, though I think it might lead to some confusion since it needs to overwrite the transform.